### PR TITLE
Add signature grinding for ECDSA signatures

### PR DESCRIPTION
**Pull request by vladimirfomene**
### Description

This PR adds a new field called `allow_grinding`
in the Signer's `SignOptions` struct that is used
to determine whether or not to grind an ECDSA signature during the signing process.

### Changelog notice

Breaking change: the BDK Signer now produces low-R signatures by default, saving one byte. If you want to preserve the original behavior, set `allow_grinding` in the `SignOptions` to `false`.

### Notes to the reviewers

This PR resolves issue #695 

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

File: src/wallet/signer.rs [hunk 1]
```diff
@@ -718,6 +724,11 @@ pub struct SignOptions {
     ///
     /// Defaults to `true`, i.e., we always try to sign with the taproot internal key.
     pub sign_with_tap_internal_key: bool,
+
+    /// Whether we should grind ECDSA signature to ensure signing with low r as bitcoin core
```
Comment (danielabrozzoni) [inline src/wallet/signer.rs]: nit: I'd remove the "as bitcoin core" part

File: src/wallet/mod.rs [hunk 2]
```diff
@@ -5546,6 +5548,35 @@ pub(crate) mod test {
...
+                    remove_partial_sigs: false,
+                    allow_grinding: true,
+                    ..Default::default()
+                },
+            )
+            .unwrap();
+
+        let key = psbt.inputs[0].partial_sigs.keys().next().unwrap();
+        let sig_len = psbt.inputs[0].partial_sigs[key].sig.serialize_der().len();
+        assert!(sig_len < 71);
```
Comment (danielabrozzoni) [inline src/wallet/mod.rs]: nit: maybe assert that it's exactly 70?

**Review by danielabrozzoni — COMMENTED** [review]
Concept ACK, the code looks good to me!

File: src/wallet/mod.rs [hunk 3]
```diff
@@ -5546,6 +5548,35 @@ pub(crate) mod test {
         assert_fee_rate!(psbt, details.fee.unwrap_or(0), fee_rate);
     }
 
+    #[test]
+    fn test_fee_rate_sign_grinding_low_r() {
+        let (wallet, _, _) = get_funded_wallet("wpkh(tprv8ZgxMBicQKsPd3EupYiPRhaMooHKUHJxNsTfYuScep13go8QFfHdtkG9nRkFGb7busX4isf6X9dURGCoKgitaApQ6MupRhZMcELAxTBRJgS/*)");
```
Comment (rajarshimaitra) [inline src/wallet/mod.rs:5557]: A description of the goal of the test like the above one would be helpful..

**Review by rajarshimaitra — COMMENTED** [review]
ACK.. Code LGTM..

File: src/wallet/mod.rs (hunk 2 above)
Comment (vladimirfomene) [reply src/wallet/mod.rs]: Makes sense

**Review by rajarshimaitra — APPROVED** [review]
ACK 68dd6d20311b28f5c8e8c9657dce8cddae4f7aa3

**Comment by danielabrozzoni** [comment]
ACK 68dd6d20311b28f5c8e8c9657dce8cddae4f7aa3

**Pull request merged**